### PR TITLE
Contrastive Loss added, First Draft Done

### PR DESCRIPTION
**Pull request by finalelement**
Signed-off-by: vnath <vnath@nvidia.com>

### Description
Adding Contrastive Loss first draft.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by finalelement** [comment]
@wyli Need your help with this specific test PT19+CUDA114.

**Comment by finalelement** [comment]
The PT19+CUDA114 is failing because of the nightly build version being used it has a known aggressive test for cosine similarity which causes the failure. For now, it has been tested on pytorch21.10-py3 docker, where it does work out.

File: monai/losses/contrastive.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        sim_ij = torch.diag(sim_matrix, self.batch_size)
+        sim_ji = torch.diag(sim_matrix, -self.batch_size)
+
+        positives = torch.cat([sim_ij, sim_ji], dim=0)
+
+        nominator = torch.exp(positives / self.temperature)
+        denominator = negatives_mask * torch.exp(sim_matrix / self.temperature)
+
+        loss_partial = -torch.log(nominator / torch.sum(denominator, dim=1))
```
Comment (wyli) [inline monai/losses/contrastive.py:86]: Perhaps the pattern of `torch.log(torch.exp(x))` could be simplified for better stability? this might be helpful https://github.com/google-research/simclr/blob/master/tf2/objective.py#L83

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me. this one requires https://github.com/Project-MONAI/MONAI/pull/3297
